### PR TITLE
Server-side CHUNKING and BINARYMIME support

### PR DESCRIPTION
**Pull request by foxcpp**
See #81. This issue is not closed since it also references client support.

### CODE REVIEW

File: chunk_reader.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+// chunkReader implements io.Reader by consuming a sequence of size-limited
+// "chunks" from underlying Reader allowing them to be interleaved with other
+// protocol data.
+//
+// It is caller responsibility to not use Reader while chunk is being processed.
+// This can be enforced by blocking on chunkEnd channel that is used to signal the
+// end of another chunk being reached.
+type chunkReader struct {
+	remainingBytes int
+	conn           io.Reader
```
Comment (corny) [inline chunk_reader.go:22]: What do you thing about renaming `conn` to `r`, `reader`, or `input`?

Comment (foxcpp) [reply chunk_reader.go]: Done

File: chunk_reader.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	return n - offset, err
+
+}
+
+func newChunkReader(conn io.Reader, maxBytes int) *chunkReader {
+	return &chunkReader{
+		remainingBytes: maxBytes,
+		r:              conn,
+		chunks:         make(chan int, 1),
+		// buffer to make sure abort() will not block if Read is not running.
```
Comment (corny) [inline chunk_reader.go:124]: abort() will never block because it just closes the channels.
Channels that are never written to don't need a buffer.

Comment (foxcpp) [reply chunk_reader.go]: Removed comment in e797215553b133513fa3c7bd3a43bd9f64699621.

**Comment by foxcpp** [comment]
> Instead of this `chunkReader` approach, can't we read data directly from `handleBdat` and forward it to the backend via an `io.Pipe`?

I wanted to preserve ability for backend to return errors without consuming the whole message but I will look if this is easily possible with io.Pipe.

**Comment by emersion** [comment]
>I wanted to preserve ability for backend to return errors without consuming the whole message but I will look if this is easily possible with io.Pipe.

`io.PipeReader.CloseWithError` could help.

**Pull request closed without merging**